src/errors.ts

```diff
@@ -0,0 +1,27 @@
+export class AuthenticationError extends Error {
+  public status = 401;
+  public type = 'authentication';
+  constructor(public message: string) {
+    super(message);
+  }
+}
+
+export class AuthorizationError extends Error {
+  public status = 403;
+  public type = 'authorization';
+  constructor(public message: string) {
+    super(message);
+  }
+}
+
+export const ERROR_CODE_STATUS_MAP = {
+  PARSE: 400,
+  VALIDATION: 422,
+  NOT_FOUND: 404,
+  INVALID_COOKIE_SIGNATURE: 401,
+  AUTHENTICATION: 401,
+  AUTHORIZATION: 403,
+  INTERNAL_SERVER_ERROR: 500,
+  UNKNOWN: 500,
+  DEFAULT: 500,
+};
```

Comment:
I was thinking about it too, it's just super ugly to do a switch for a simple hardcoded mapping without any logic in it.

What do you think about implementing a default map? E.g.
```
export const DEFAULT = Symbol();

export class MapWithDefault<K, V> extends Map<K | typeof DEFAULT, V> {
    get(key: K): V { return super.get(key) ?? (super.get(DEFAULT) as V); }
}

const map = new MapWithDefault<string, number >([
    ['PARSE', 400],
    ['VALIDATION', 422],
    ['NOT_FOUND', 404],
    ['INVALID_COOKIE_SIGNATURE', 401],
    ['AUTHENTICATION', 401],
    ['AUTHORIZATION', 403],
    ['INTERNAL_SERVER_ERROR', 500],
    ['UNKNOWN', 500],
    [DEFAULT, 500],
]);

console.log(map.get('PARSE')); 
console.log(map.get('AUTHENTICATION'));
console.log(map.get('SOMETHING_RANDOM'));
```

Reply:
Man... I was thinking the same thing! I even had almost the same code in my IDE. 😍

I think I went with switch-case because I was splitting hairs about adding yet another entity to the codebase, but I do think it's cleaner.

Continuing on that, you've probably also thought of [defaultdict](https://docs.python.org/3/library/collections.html#defaultdict-objects) from python collections. It is a great little class that I have used in many of my leetcode sessions 😀 That one takes a factory instead of a constant, 

There is something similar in TypeScript landscape with [Mnemonist](https://yomguithereal.github.io/mnemonist/default-map). It seems to be [mature and popular enough](https://github.com/Yomguithereal/mnemonist), with a solid 56:2.1k issue:star ratio on GitHub.

Reply:
I'm not even surprised you thought the same 😁

Yea, pythons' defaultdict (and generally the collections module) is really nice! This mnemonist package seems really interesting, although I might not install it for only a default dict. It's probably better to agree on an implementation and just have it in our codebase.

After [this](https://www.davidhaney.io/npm-left-pad-have-we-forgotten-how-to-program/) incident I prefer not to have too many dependencies 😅

Reply:
Haha, wow. What an interesting read, if a bit too gate-keepy for my taste. I find it toxic of programmers to tell other programmers they're not 'ReAl PrOgRaMmErS'. There are many types of programmers. When studying for interviews, I challenged myself to solve dozens of leetcode questions ranked by hard-to-easy under a time pressure and with the restraint of no Googling. I would still not call anyone who doesn't code their own utils in 5 minutes someone who "doesn't know how to code"; I especially dislike this paragraph:

> Stringing APIs together and calling it programming doesn’t make it programming. It’s some crazy form of dependency hacking that involves the cloud, over-engineering things, and complexity far beyond what’s actually needed to create great applications.

How dramatic! In production, you do whatever it takes to deliver good code, including tapping into the community of great, battle-tested packages. You do it to save time and justify your choices more easily. 

Ultimately, this is just debating in extremes. Just like a lot of these things, the answer probably lies somewhere in the middle. We shouldn't use packages for everything, and you're right, we shouldn't use one here. But let's try to be positive and focus on education rather than gatekeeping. So far, we're doing great.